File: day2.py
data = open("data/day2.txt", 'r').read().splitlines()


def calculate(instrs):
    horiz = 0
    depth = 0
    for i in instrs:
        tup = i.split(" ")
        direction = tup[0]
        amount = int(tup[1])
        if direction == "down":
            depth += amount
        elif direction == "up":
            depth -= amount
        elif direction == "forward":
            horiz += amount
    return (horiz, depth)


def part2(instrs):
    aim = 0
    depth = 0
    horiz = 0
    for i in instrs:
        tup = i.split(" ")
        direction = tup[0]
        amount = int(tup[1])
        if direction == "down":
            aim += amount
        elif direction == "up":
            aim -= amount
        elif direction == "forward":
            horiz += amount
            depth += aim * amount
    return (horiz, depth)


res = calculate(data)
print(res[0] * res[1])
res = part2(data)
print(res[0] * res[1])

File: src/day02/day2.kt
// kotlinc day2.kt -include-runtime -d day2.jar

import java.io.File;

fun main() {
    println("Hello, world!")
    val input = File("input.txt").useLines { it.toList() }
    val instrs = input.map { parse(it) }
    println(part1(instrs))
    println(part2(instrs))
}

fun part1(instrs: List<Pair<String, Int>>): Int {
    var depth = 0;
    var horiz = 0;
    instrs.forEach {
        when (it.first) {
            "forward" -> horiz += it.second
            "down" -> depth += it.second
            "up" -> depth -= it.second
            else -> {}
        }
    }
    return depth * horiz;
}

fun part2(instrs: List<Pair<String, Int>>): Int {
    var depth = 0;
    var horiz = 0;
    var aim = 0;
    instrs.forEach {
        when (it.first) {
            "forward" -> { 
                horiz += it.second
                depth += (aim * it.second)
            }
            "down" -> aim += it.second
            "up" -> aim -= it.second
            else -> {}
        }
    }
    return depth * horiz;
}

fun parse(line: String): Pair<String, Int> {
    val info = line.split(" ")
    return Pair(info[0], info[1].toInt())
}